src/com/unbounce/treajure/concurrent/mvar.clj

```diff
@@ -0,0 +1,81 @@
+(ns com.unbounce.treajure.concurrent.mvar
+  "Haskell's MVar abstraction in Clojure.
+
+  Check: http://chimera.labs.oreilly.com/books/1230000000929/ch07.html for more
+  info around MVars."
+  (:import [java.util.concurrent Semaphore]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Lightweight Performance mutable tuple implementation in Clojure
+;; (Look ma! no atoms!)
+
+(defprotocol IMCons
+  (mcar [p])
+  (mcdr [p])
+  (set-mcar! [p val])
+  (set-mcdr! [p val]))
+
+(deftype MCons [^{:volatile-mutable true} car
+                ^{:volatile-mutable true} cdr]
+  IMCons
+  (mcar [this] car)
+  (mcdr [this] cdr)
+  (set-mcar! [this val]
+    (set! car val))
+  (set-mcdr! [this val]
+    (set! cdr val)))
+
+(defn mcons [a b]
+  (MCons. a b))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; MVar abstraction
+
+(declare new-empty-mvar new-mvar put-mvar take-mvar read-mvar)
+
+(defn new-empty-mvar
+  "Create an empty MVar"
+  []
+  (let [take-sem (Semaphore. 1)
+        put-sem  (Semaphore. 1)]
```

Comment:
I know we talked about the `Exchanger` API yesterday and it's not super pretty. However, I do think it might simplify this code quite a bit.

@jpb, see: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Exchanger.html